scripts/zones/Nashmau/npcs/Kilusha.lua

```diff
@@ -1,21 +1,107 @@
 -----------------------------------
 -- Area: Nashmau
 --  NPC: Kilusha
--- Standard Info NPC
+-- Einherjar-related NPC, Smoldering Glass, Therion Ichor items
+-- !pos 0.373 -6.667 14.712 53
 -----------------------------------
+local ID = require("scripts/zones/Nashmau/IDs")
+require("scripts/globals/items")
+require("scripts/globals/keyitems")
+require("scripts/globals/missions")
+require("scripts/globals/npc_util")
+-----------------------------------
+
 local entity = {}
 
 entity.onTrade = function(player, npc, trade)
+    local lampCost = 60000 -- base cost without RHAPSODY_IN_AZURE key item
+
+    if player:hasKeyItem(xi.ki.RHAPSODY_IN_AZURE) then
+        lampCost = 1000
+    end
+
+    if npcUtil.tradeHasExactly(trade, {{"gil", lampCost}}) and player:getCharVar("EinherjarIntro") ~= 1 then
+        if npcUtil.giveItem(player, xi.items.SMOLDERING_LAMP) then
+            player:tradeComplete()
+            player:startEvent(25)
+        else
+            player:messageSpecial( ID.text.ITEM_CANNOT_BE_OBTAINED, xi.items.SMOLDERING_LAMP)
+        end
+    end
 end
 
 entity.onTrigger = function(player, npc)
-    player:startEvent(22)
+    local ichor = player:getCurrency("therion_ichor")
+    local allowValkyrieBuying = 29360128 -- set this to 0 if you wish to allow players to buy feather key items without KI
+    local lampCost = 60000 -- base cost without RHAPSODY_IN_AZURE key item
+    local reentryTime = 20 -- in hours
+    local toau = player:hasCompletedMission(xi.mission.log_id.TOAU, xi.mission.id.toau.IMMORTAL_SENTRIES)
+
+    if player:hasKeyItem(xi.ki.RHAPSODY_IN_AZURE) then
+        lampCost = 1000
+        reentryTime = 1
+    end
+
+    if player:hasKeyItem(xi.ki.MARK_OF_THE_EINHERJAR) then -- June 2012 update added Valkyrie items
+        allowValkyrieBuying = 0
+    end
+
+    if player:getMainLvl() <= 59 or not toau then
+        player:startEvent(22) -- worthless CS
+    elseif (player:getMainLvl() >= 60 or toau) and player:getCharVar("EinherjarIntro") == 1 then
+        player:startEvent(23, lampCost, 856, 3, 616, 10, 172, 172, 0) -- Einherjar introduction
+    else
+        player:startEvent(24, lampCost, 856, 3, reentryTime, 10, 135, allowValkyrieBuying, ichor)
+        player:setLocalVar("reentryTime", reentryTime)
+    end
 end
 
 entity.onEventUpdate = function(player, csid, option)
+    if csid == 24 and option == 6 then -- about Entry Conditions
+        player:updateEvent(53,10,3,player:getLocalVar("reentryTime"),10,231,xi.items.SMOLDERING_LAMP,xi.items.GLOWING_LAMP)
+    end
 end
 
 entity.onEventFinish = function(player, csid, option)
+    if csid == 23 then
+        player:setCharVar("EinherjarIntro", 0) -- deletes CharVar set at character creation
+    elseif csid == 24 and option ~= 1073741824 and option ~= 0 then
+        local kilushaItems = {
```

Comment:
I would probably define this table outside of the function to make it a bit more simple to understand, but either way works.